### PR TITLE
[WIP] Add TEDLIUM dataset

### PR DESCRIPTION
**Pull request by sanchit-gandhi**
Adds the TED-LIUM dataset https://www.tensorflow.org/datasets/catalog/tedlium#tedliumrelease3 

TODO:

- [x] Port `tedium.py` from TF datasets using `convert_dataset.sh` script
- [x] Make `load_dataset` work
- [ ] ~~Run `datasets-cli` command to generate `dataset_infos.json`~~
- [ ] ~~Create dummy data for continuous testing~~
- [ ] ~~Dummy data tests~~
- [ ] ~~Real data tests~~
- [ ] Create the metadata JSON
- [ ] Close PR and add directly to the Hub under LIUM org

### CODE REVIEW

**Comment by sanchit-gandhi** [comment]
```python
from datasets import load_dataset

ds = load_dataset('./datasets/tedlium', 'release1', cache_dir='/home/sanchitgandhi/cache')
```

```
Downloading and preparing dataset tedlium/release1 to /home/sanchitgandhi/cache/tedlium/release1/1.0.1/5a9fcb97b4b52d5a1c9dc7bde4b1d5994cd89c4a3425ea36c789bf6096fee4f0...
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/sanchit_huggingface_co/datasets/src/datasets/load.py", line 1703, in load_dataset
    builder_instance.download_and_prepare(
  File "/home/sanchit_huggingface_co/datasets/src/datasets/builder.py", line 605, in download_and_prepare
    self._download_and_prepare(
  File "/home/sanchit_huggingface_co/datasets/src/datasets/builder.py", line 1240, in _download_and_prepare
    raise MissingBeamOptions(
datasets.builder.MissingBeamOptions: Trying to generate a dataset using Apache Beam, yet no Beam Runner or PipelineOptions() has been provided in `load_dataset` or in the builder arguments. For big datasets it has to run on large-scale data processing tools like Dataflow, Spark, etc. More information about Apache Beam runners at https://beam.apache.org/documentation/runners/capability-matrix/
If you really want to run it locally because you feel like the Dataset is small enough, you can use the local beam runner called `DirectRunner` (you may run out of memory). 
Example of usage: 
        `load_dataset('tedlium', 'release1', beam_runner='DirectRunner')`
```
Specifying the `beam_runner='DirectRunner'` works:
```python
from datasets import load_dataset

ds = load_dataset('./datasets/tedlium', 'release1', cache_dir='home/sanchitgandhi/cache', beam_runner='DirectRunner')
```

**Comment by sanchit-gandhi** [comment]
Extra Python imports/Linux packages:
```
pip install pydub
sudo apt install ffmpeg
```

**Comment by sanchit-gandhi** [comment]
Heavily inspired by the TF datasets script at: https://github.com/tensorflow/datasets/blob/master/tensorflow_datasets/audio/tedlium.py

The TF datasets script uses the module AudioSegment from the package `pydub` (https://github.com/jiaaro/pydub), which is used to to open the audio files (stored in .sph format):
https://github.com/huggingface/datasets/blob/61bf6123634bf6e7c7287cd6097909eb26118c58/datasets/tedlium/tedlium.py#L167-L170
This package requires the pip install of `pydub` and the system installation of `ffmpeg`: https://github.com/jiaaro/pydub#installation
Is it ok to use this import? Or do we tend to avoid introducing additional dependencies?

The TF datasets script also uses `_build_pcollection`:
https://github.com/huggingface/datasets/blob/8afbbb6fe66b40d05574e2e72e65e974c72ae769/datasets/tedlium/tedlium.py#L200-L206
However, I was advised against using `beam` logic. Thus, I have reverted to generating the examples file-by-file: https://github.com/huggingface/datasets/blob/61bf6123634bf6e7c7287cd6097909eb26118c58/datasets/tedlium/tedlium.py#L112-L138

I am now able to generate examples by running the `load_dataset` command:
```python
from datasets import load_dataset

ds = load_dataset('./datasets/tedlium', 'release1', cache_dir='/home/sanchitgandhi/cache')
```

Generating the examples is **extremely** slow: it takes ~1 second per example, so ~60k seconds for the train set. Is there a way of paralleling this to make it faster?

**Comment by lhoestq** [comment]
> This package requires the pip install of pydub and the system installation of ffmpeg: https://github.com/jiaaro/pydub#installation
Is it ok to use these packages? Or do we tend to avoid introducing additional dependencies?

It's ok, windows users will have have a bad time but I'm not sure we can do much about it.

> Here, generating examples is extremely slow: it takes ~1 second per example, so ~60k seconds for the train set (~16 hours). Is there a way of paralleling this to make it faster?

Not at the moment. For such cases we advise hosting the dataset ourselves in a processed format. The license doesn't allow this since the license is "NoDerivatives". Currently the only way to parallelize it is by keeping is as a beam dataset and let users pay Google Dataflow to process it (or use spark or whatever).

**Comment by mariosasko** [comment]
> This package requires the pip install of pydub and the system installation of ffmpeg: https://github.com/jiaaro/pydub#installation
Is it ok to use these packages? Or do we tend to avoid introducing additional dependencies?

`soundfile`, which is a required audio dependency, should also work with `.sph` files, no?

**Comment by sanchit-gandhi** [comment]
> `soundfile`, which is a required audio dependency, should also work with `.sph` files, no?

Awesome, thanks for the pointer @mariosasko! Switched `pydub` to `soundfile`, and having specifying the `dtype` argument in `soundfile.read` as `np.int16`, the arrays match with those from `pydub` ✅

I also did some heavy optimising of the script with the processing of the `.stm` and `.sph` files - it now runs 2000x faster than before, so there probably isn't a need to upload the data to the Hub @lhoestq. The total processing time is just ~2mins now 🚀

File: datasets/tedlium/tedlium.py [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+            features=features,
+            supervised_keys=("audio", "text"),
+            homepage=self.config.url,
+            license=_LICENSE,
+            citation=self.config.citation,
+            task_templates=[AutomaticSpeechRecognition(audio_column="audio", transcription_column="text")],
+        )
+
+    def _split_generators(self, dl_manager):
+        data_dir = dl_manager.download_and_extract(self.config.download_url)
```
Comment (lhoestq) [inline datasets/tedlium/tedlium.py:195]: To make the dataset streaming work with TAR archives, you have to use `dl_manager.iter_archive`, not `download_and_extract`, see librispeech_asr for example:

https://github.com/huggingface/datasets/blob/095d12ff7414df118f60e00cd6494299a881743a/datasets/librispeech_asr/librispeech_asr.py#L116-L130

File: datasets/tedlium/README.md [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+multilinguality:
+- monolingual
+size_categories:
+- 10K<n<100K
+source_datasets:
+- original
+task_categories:
+- automatic-speech-recognition
+task_ids:
+- automatic-speech-recognition
```
Comment (lhoestq) [inline datasets/tedlium/README.md:20]: no need to be redundant with `automatic-speech-recognition` ;)
```suggestion
task_ids: []
```

**Review by lhoestq — COMMENTED** [review]
That's super good news ! amazing :)

I added a comment in case you wanted to also make the dataset work in streaming mode.

Also you can add dummy data at `datasets/tedlium/dummy/release1/1.0.1/dummy_data.zip` if you'd like to have the dataset tested in the CI.

It must contain a directory named after the TAR archive that has to be downloaded (including the ".tgz" suffix), and in this directory you can place two or three dummy audio files (using the same name or structure as the original TAR archive).

File: datasets/tedlium/tedlium.py (hunk 1 above)
Comment (sanchit-gandhi) [reply datasets/tedlium/tedlium.py]: Hey @lhoestq, I making good progress! The issue I’m facing is that the data for all three splits (train, dev and test) is downloaded from one `.tar` file at http://www.openslr.org/resources/19/TEDLIUM_release2.tar.gz. The subsets are separated into sub-directories when the file is extracted (`TEDLIUM_release1/train`, `TEDLIUM_release1/dev`, `TEDLIUM_release1/test`). Thus, in defining `archive_path` as the `dl_manger.download`, the variable points to **all three subsets**, and the iterator `filepath` a generator object over the data for all three subsets: 
https://github.com/huggingface/datasets/blob/18ff32a52a857af4cfa997d9e192844478e11531/datasets/tedlium/tedlium.py#L194-L208
This means that in `_generate_examples`, the iterator `filepath` has to go through every example and check if it belongs to the correct split:
https://github.com/huggingface/datasets/blob/18ff32a52a857af4cfa997d9e192844478e11531/datasets/tedlium/tedlium.py#L246-L249
Does the `dl_manager` have functionality to produce an iterator for a specific subset of the data extracted from the `.tar` file?

Comment (lhoestq) [reply datasets/tedlium/tedlium.py]: Feel free to host the data on the Hub and separate the splits in different files :)

**Comment by sanchit-gandhi** [comment]
TEDLIUM completed and uploaded to the HF Hub: https://huggingface.co/datasets/LIUM/tedlium

**Comment by lhoestq** [comment]
Awesome !

**Pull request closed without merging**